### ISSUE TITLE
Copying product with tags generates error on subsequent edits

### ISSUE THREAD

**Issue by EddieGreen** [nopSolutions/nopCommerce#3040]
When a product with tags is copied, further editing generates an error (see log below) and returns a blank screen.

Removing the tags from the copied product (not tried removing from original) clears the error. Re-adding the same tags manually does not recreate an error.

nopCommerce version: 4.10

Log entries:
```
info: Microsoft.EntityFrameworkCore.Database.Command[20101]
      Executed DbCommand (17ms) [Parameters=[@p103='?' (DbType = Int32), @p0='?' (DbType = Decimal), @p1='?' (Size = 4000), @p2='?' (DbType = Boolean), @p3='?' (DbType = Boolean), @p4='?' (DbType = Boolean), @p5='?' (Size = 1000), @p6='?' (DbType = Int32), @p7='?' (DbType = Int32), @p8='?' (DbType = Boolean), @p9='?' (DbType = DateTime2), @p10='?' (DbType = Boolean), @p11='?' (DbType = DateTime2), @p12='?' (DbType = Int32), @p13='?' (DbType = Decimal), @p14='?' (DbType = Decimal), @p15='?' (DbType = Int32), @p16='?' (DbType = Boolean), @p17='?' (DbType = Int32), @p18='?' (DbType = Boolean), @p19='?' (DbType = DateTime2), @p20='?' (DbType = Boolean), @p21='?' (DbType = Boolean), @p22='?' (DbType = Int32), @p23='?' (DbType = Boolean), @p24='?' (DbType = Boolean), @p25='?' (DbType = Int32), @p26='?' (DbType = Boolean), @p27='?' (DbType = Boolean), @p28='?' (DbType = Int32), @p29='?' (DbType = Int32), @p30='?' (DbType = Int32), @p31='?' (Size = 4000), @p32='?' (DbType = Int32), @p33='?' (Size = 400), @p34='?' (DbType = Boolean), @p35='?' (DbType = Boolean), @p36='?' (DbType = Boolean), @p37='?' (DbType = Boolean), @p38='?' (DbType = Decimal), @p39='?' (DbType = Boolean), @p40='?' (DbType = Boolean), @p41='?' (DbType = Boolean), @p42='?' (DbType = Boolean), @p43='?' (DbType = Boolean), @p44='?' (DbType = Boolean), @p45='?' (DbType = Boolean), @p46='?' (DbType = Boolean), @p47='?' (DbType = Decimal), @p48='?' (DbType = Boolean), @p49='?' (DbType = Int32), @p50='?' (DbType = Int32), @p51='?' (Size = 400), @p52='?' (DbType = Boolean), @p53='?' (DbType = DateTime2), @p54='?' (DbType = DateTime2), @p55='?' (DbType = Int32), @p56='?' (DbType = Decimal), @p57='?' (Size = 4000), @p58='?' (Size = 400), @p59='?' (Size = 400), @p60='?' (DbType = Int32), @p61='?' (DbType = Decimal), @p62='?' (Size = 400), @p63='?' (DbType = Int32), @p64='?' (DbType = Int32), @p65='?' (DbType = Boolean), @p66='?' (DbType = Int32), @p67='?' (DbType = Decimal), @p68='?' (DbType = Int32), @p69='?' (DbType = Int32), @p70='?', @p71='?' (DbType = Int32), @p72='?' (DbType = DateTime2), @p73='?' (DbType = Decimal), @p74='?' (DbType = Int32), @p75='?' (DbType = Decimal), @p76='?' (DbType = Int32), @p77='?' (DbType = Int32), @p78='?' (DbType = Boolean), @p79='?' (DbType = Int32), @p80='?' (DbType = Int32), @p81='?' (DbType = Int32), @p82='?' (DbType = Int32), @p83='?' (DbType = Int32), @p84='?' (DbType = Boolean), @p85='?' (Size = 1000), @p86='?' (DbType = Int32), @p87='?' (DbType = Boolean), @p88='?' (Size = 4000), @p89='?' (DbType = Boolean), @p90='?' (Size = 400), @p91='?' (DbType = Int32), @p92='?' (DbType = Boolean), @p93='?' (DbType = Int32), @p94='?' (DbType = Boolean), @p95='?' (DbType = DateTime2), @p96='?' (DbType = Boolean), @p97='?' (Size = 4000), @p98='?' (DbType = Int32), @p99='?' (DbType = Boolean), @p100='?' (DbType = Int32), @p101='?' (DbType = Decimal), @p102='?' (DbType = Decimal)], CommandType='Text', CommandTimeout='30']
      SET NOCOUNT ON;
      UPDATE [Product] SET [AdditionalShippingCharge] = @p0, [AdminComment] = @p1, [AllowAddingOnlyExistingAttributeCombinations] = @p2, [AllowBackInStockSubscriptions] = @p3, [AllowCustomerReviews] = @p4, [AllowedQuantities] = @p5, [ApprovedRatingSum] = @p6, [ApprovedTotalReviews] = @p7, [AutomaticallyAddRequiredProducts] = @p8, [AvailableEndDateTimeUtc] = @p9, [AvailableForPreOrder] = @p10, [AvailableStartDateTimeUtc] = @p11, [BackorderModeId] = @p12, [BasepriceAmount] = @p13, [BasepriceBaseAmount] = @p14, [BasepriceBaseUnitId] = @p15, [BasepriceEnabled] = @p16, [BasepriceUnitId] = @p17, [CallForPrice] = @p18, [CreatedOnUtc] = @p19, [CustomerEntersPrice] = @p20, [Deleted] = @p21, [DeliveryDateId] = @p22, [DisableBuyButton] = @p23, [DisableWishlistButton] = @p24, [DisplayOrder] = @p25, [DisplayStockAvailability] = @p26, [DisplayStockQuantity] = @p27, [DownloadActivationTypeId] = @p28, [DownloadExpirationDays] = @p29, [DownloadId] = @p30, [FullDescription] = @p31, [GiftCardTypeId] = @p32, [Gtin] = @p33, [HasDiscountsApplied] = @p34, [HasSampleDownload] = @p35, [HasTierPrices] = @p36, [HasUserAgreement] = @p37, [Height] = @p38, [IsDownload] = @p39, [IsFreeShipping] = @p40, [IsGiftCard] = @p41, [IsRecurring] = @p42, [IsRental] = @p43, [IsShipEnabled] = @p44, [IsTaxExempt] = @p45, [IsTelecommunicationsOrBroadcastingOrElectronicServices] = @p46, [Length] = @p47, [LimitedToStores] = @p48, [LowStockActivityId] = @p49, [ManageInventoryMethodId] = @p50, [ManufacturerPartNumber] = @p51, [MarkAsNew] = @p52, [MarkAsNewEndDateTimeUtc] = @p53, [MarkAsNewStartDateTimeUtc] = @p54, [MaxNumberOfDownloads] = @p55, [MaximumCustomerEnteredPrice] = @p56, [MetaDescription] = @p57, [MetaKeywords] = @p58, [MetaTitle] = @p59, [MinStockQuantity] = @p60, [MinimumCustomerEnteredPrice] = @p61, [Name] = @p62, [NotApprovedRatingSum] = @p63, [NotApprovedTotalReviews] = @p64, [NotReturnable] = @p65, [NotifyAdminForQuantityBelow] = @p66, [OldPrice] = @p67, [OrderMaximumQuantity] = @p68, [OrderMinimumQuantity] = @p69, [OverriddenGiftCardAmount] = @p70, [ParentGroupedProductId] = @p71, [PreOrderAvailabilityStartDateTimeUtc] = @p72, [Price] = @p73, [ProductAvailabilityRangeId] = @p74, [ProductCost] = @p75, [ProductTemplateId] = @p76, [ProductTypeId] = @p77, [Published] = @p78, [RecurringCycleLength] = @p79, [RecurringCyclePeriodId] = @p80, [RecurringTotalCycles] = @p81, [RentalPriceLength] = @p82, [RentalPricePeriodId] = @p83, [RequireOtherProducts] = @p84, [RequiredProductIds] = @p85, [SampleDownloadId] = @p86, [ShipSeparately] = @p87, [ShortDescription] = @p88, [ShowOnHomePage] = @p89, [Sku] = @p90, [StockQuantity] = @p91, [SubjectToAcl] = @p92, [TaxCategoryId] = @p93, [UnlimitedDownloads] = @p94, [UpdatedOnUtc] = @p95, [UseMultipleWarehouses] = @p96, [UserAgreementText] = @p97, [VendorId] = @p98, [VisibleIndividually] = @p99, [WarehouseId] = @p100, [Weight] = @p101, [Width] = @p102
      WHERE [Id] = @p103;
      SELECT @@ROWCOUNT;
fail: Microsoft.EntityFrameworkCore.Database.Command[20102]
      Failed executing DbCommand (14ms) [Parameters=[@p0='?' (DbType = Int32), @p1='?' (DbType = Int32)], CommandType='Text', CommandTimeout='30']
      SET NOCOUNT ON;
      INSERT INTO [Product_ProductTag_Mapping] ([Product_Id], [ProductTag_Id])
      VALUES (@p0, @p1);
System.Data.SqlClient.SqlException (0x80131904): Violation of PRIMARY KEY constraint 'PK_Product_ProductTag_Mapping'. Cannot insert duplicate key in object 'dbo.Product_ProductTag_Mapping'. The duplicate key value is (3, 1).
The statement has been terminated.
   at System.Data.SqlClient.SqlConnection.OnError(SqlException exception, Boolean breakConnection, Action`1 wrapCloseInAction)
   at System.Data.SqlClient.TdsParser.ThrowExceptionAndWarning(TdsParserStateObject stateObj, Boolean callerHasConnectionLock, Boolean asyncClose)
   at System.Data.SqlClient.TdsParser.TryRun(RunBehavior runBehavior, SqlCommand cmdHandler, SqlDataReader dataStream, BulkCopySimpleResultSet bulkCopyHandler, TdsParserStateObject stateObj, Boolean& dataReady)
   at System.Data.SqlClient.SqlDataReader.TryConsumeMetaData()
   at System.Data.SqlClient.SqlDataReader.get_MetaData()
   at System.Data.SqlClient.SqlCommand.FinishExecuteReader(SqlDataReader ds, RunBehavior runBehavior, String resetOptionsString)
   at System.Data.SqlClient.SqlCommand.RunExecuteReaderTds(CommandBehavior cmdBehavior, RunBehavior runBehavior, Boolean returnStream, Boolean async, Int32 timeout, Task& task, Boolean asyncWrite, SqlDataReader ds)
   at System.Data.SqlClient.SqlCommand.ExecuteReader(CommandBehavior behavior)
   at Microsoft.EntityFrameworkCore.Storage.Internal.RelationalCommand.Execute(IRelationalConnection connection, DbCommandMethod executeMethod, IReadOnlyDictionary`2 parameterValues)
ClientConnectionId:a5f3c5db-cd0e-4245-8e29-b913f0f3f138
Error Number:2627,State:1,Class:14
fail: Microsoft.EntityFrameworkCore.Update[10000]
      An exception occurred in the database while saving changes for context type 'Nop.Data.NopObjectContext'.
      Microsoft.EntityFrameworkCore.DbUpdateException: An error occurred while updating the entries. See the inner exception for details. ---> System.Data.SqlClient.SqlException: Violation of PRIMARY KEY constraint 'PK_Product_ProductTag_Mapping'. Cannot insert duplicate key in object 'dbo.Product_ProductTag_Mapping'. The duplicate key value is (3, 1).
      The statement has been terminated.
         at System.Data.SqlClient.SqlConnection.OnError(SqlException exception, Boolean breakConnection, Action`1 wrapCloseInAction)
         at System.Data.SqlClient.TdsParser.ThrowExceptionAndWarning(TdsParserStateObject stateObj, Boolean callerHasConnectionLock, Boolean asyncClose)
         at System.Data.SqlClient.TdsParser.TryRun(RunBehavior runBehavior, SqlCommand cmdHandler, SqlDataReader dataStream, BulkCopySimpleResultSet bulkCopyHandler, TdsParserStateObject stateObj, Boolean& dataReady)
         at System.Data.SqlClient.SqlDataReader.TryConsumeMetaData()
         at System.Data.SqlClient.SqlDataReader.get_MetaData()
         at System.Data.SqlClient.SqlCommand.FinishExecuteReader(SqlDataReader ds, RunBehavior runBehavior, String resetOptionsString)
         at System.Data.SqlClient.SqlCommand.RunExecuteReaderTds(CommandBehavior cmdBehavior, RunBehavior runBehavior, Boolean returnStream, Boolean async, Int32 timeout, Task& task, Boolean asyncWrite, SqlDataReader ds)
         at System.Data.SqlClient.SqlCommand.ExecuteReader(CommandBehavior behavior)
         at Microsoft.EntityFrameworkCore.Storage.Internal.RelationalCommand.Execute(IRelationalConnection connection, DbCommandMethod executeMethod, IReadOnlyDictionary`2 parameterValues)
         at Microsoft.EntityFrameworkCore.Storage.Internal.RelationalCommand.ExecuteReader(IRelationalConnection connection, IReadOnlyDictionary`2 parameterValues)
         at Microsoft.EntityFrameworkCore.Update.ReaderModificationCommandBatch.Execute(IRelationalConnection connection)
         --- End of inner exception stack trace ---
         at Microsoft.EntityFrameworkCore.Update.ReaderModificationCommandBatch.Execute(IRelationalConnection connection)
         at Microsoft.EntityFrameworkCore.Update.Internal.BatchExecutor.Execute(DbContext _, ValueTuple`2 parameters)
         at Microsoft.EntityFrameworkCore.SqlServer.Storage.Internal.SqlServerExecutionStrategy.Execute[TState,TResult](TState state, Func`3 operation, Func`3 verifySucceeded)
         at Microsoft.EntityFrameworkCore.Update.Internal.BatchExecutor.Execute(IEnumerable`1 commandBatches, IRelationalConnection connection)
         at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.StateManager.SaveChanges(IReadOnlyList`1 entriesToSave)
         at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.StateManager.SaveChanges(Boolean acceptAllChangesOnSuccess)
         at Microsoft.EntityFrameworkCore.DbContext.SaveChanges(Boolean acceptAllChangesOnSuccess)
Microsoft.EntityFrameworkCore.DbUpdateException: An error occurred while updating the entries. See the inner exception for details. ---> System.Data.SqlClient.SqlException: Violation of PRIMARY KEY constraint 'PK_Product_ProductTag_Mapping'. Cannot insert duplicate key in object 'dbo.Product_ProductTag_Mapping'. The duplicate key value is (3, 1).
The statement has been terminated.
   at System.Data.SqlClient.SqlConnection.OnError(SqlException exception, Boolean breakConnection, Action`1 wrapCloseInAction)
   at System.Data.SqlClient.TdsParser.ThrowExceptionAndWarning(TdsParserStateObject stateObj, Boolean callerHasConnectionLock, Boolean asyncClose)
   at System.Data.SqlClient.TdsParser.TryRun(RunBehavior runBehavior, SqlCommand cmdHandler, SqlDataReader dataStream, BulkCopySimpleResultSet bulkCopyHandler, TdsParserStateObject stateObj, Boolean& dataReady)
   at System.Data.SqlClient.SqlDataReader.TryConsumeMetaData()
   at System.Data.SqlClient.SqlDataReader.get_MetaData()
   at System.Data.SqlClient.SqlCommand.FinishExecuteReader(SqlDataReader ds, RunBehavior runBehavior, String resetOptionsString)
   at System.Data.SqlClient.SqlCommand.RunExecuteReaderTds(CommandBehavior cmdBehavior, RunBehavior runBehavior, Boolean returnStream, Boolean async, Int32 timeout, Task& task, Boolean asyncWrite, SqlDataReader ds)
   at System.Data.SqlClient.SqlCommand.ExecuteReader(CommandBehavior behavior)
   at Microsoft.EntityFrameworkCore.Storage.Internal.RelationalCommand.Execute(IRelationalConnection connection, DbCommandMethod executeMethod, IReadOnlyDictionary`2 parameterValues)
   at Microsoft.EntityFrameworkCore.Storage.Internal.RelationalCommand.ExecuteReader(IRelationalConnection connection, IReadOnlyDictionary`2 parameterValues)
   at Microsoft.EntityFrameworkCore.Update.ReaderModificationCommandBatch.Execute(IRelationalConnection connection)
   --- End of inner exception stack trace ---
   at Microsoft.EntityFrameworkCore.Update.ReaderModificationCommandBatch.Execute(IRelationalConnection connection)
   at Microsoft.EntityFrameworkCore.Update.Internal.BatchExecutor.Execute(DbContext _, ValueTuple`2 parameters)
   at Microsoft.EntityFrameworkCore.SqlServer.Storage.Internal.SqlServerExecutionStrategy.Execute[TState,TResult](TState state, Func`3 operation, Func`3 verifySucceeded)
   at Microsoft.EntityFrameworkCore.Update.Internal.BatchExecutor.Execute(IEnumerable`1 commandBatches, IRelationalConnection connection)
   at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.StateManager.SaveChanges(IReadOnlyList`1 entriesToSave)
   at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.StateManager.SaveChanges(Boolean acceptAllChangesOnSuccess)
   at Microsoft.EntityFrameworkCore.DbContext.SaveChanges(Boolean acceptAllChangesOnSuccess)
fail: Microsoft.AspNetCore.Diagnostics.ExceptionHandlerMiddleware[0]
      An exception was thrown attempting to execute the error handler.
System.InvalidOperationException: The instance of entity type 'ProductProductTagMapping' cannot be tracked because another instance with the same key value for {'ProductId', 'ProductTagId'} is already being tracked. When attaching existing entities, ensure that only one entity instance with a given key value is attached. Consider using 'DbContextOptionsBuilder.EnableSensitiveDataLogging' to see the conflicting key values.
   at Nop.Web.Framework.Infrastructure.Extensions.ApplicationBuilderExtensions.<>c.<UseNopExceptionHandler>b__1_1(HttpContext context) in D:\Data\My Documents\Visual Studio 2017\Projects\Artplinths 2018\Presentation\Nop.Web.Framework\Infrastructure\Extensions\ApplicationBuilderExtensions.cs:line 76
   at Microsoft.AspNetCore.Diagnostics.ExceptionHandlerMiddleware.<Invoke>d__6.MoveNext()
fail: Microsoft.AspNetCore.Diagnostics.ExceptionHandlerMiddleware[0]
      An unhandled exception has occurred: The instance of entity type 'ProductProductTagMapping' cannot be tracked because another instance with the same key value for {'ProductId', 'ProductTagId'} is already being tracked. When attaching existing entities, ensure that only one entity instance with a given key value is attached. Consider using 'DbContextOptionsBuilder.EnableSensitiveDataLogging' to see the conflicting key values.
System.InvalidOperationException: The instance of entity type 'ProductProductTagMapping' cannot be tracked because another instance with the same key value for {'ProductId', 'ProductTagId'} is already being tracked. When attaching existing entities, ensure that only one entity instance with a given key value is attached. Consider using 'DbContextOptionsBuilder.EnableSensitiveDataLogging' to see the conflicting key values.
   at Nop.Web.Framework.Infrastructure.Extensions.ApplicationBuilderExtensions.<>c.<UseNopExceptionHandler>b__1_1(HttpContext context) in D:\Data\My Documents\Visual Studio 2017\Projects\Artplinths 2018\Presentation\Nop.Web.Framework\Infrastructure\Extensions\ApplicationBuilderExtensions.cs:line 76
   at Microsoft.AspNetCore.Diagnostics.ExceptionHandlerMiddleware.<Invoke>d__6.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at Microsoft.AspNetCore.Diagnostics.ExceptionHandlerMiddleware.<Invoke>d__6.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.AspNetCore.Diagnostics.ExceptionHandlerMiddleware.<Invoke>d__6.MoveNext()
```

**Comment by EddieGreen** [comment]
Further to this, I am tracking down some problems with urls and examining UrlRecord table.

As background, the database was originally created with a clean 4.00 installation and subsequently upgraded with 4.10.

This is what I saw when I first looked at the table:
```
Id	EntityId	EntityName	Slug	IsActive	LanguageId
1	1	Topic	about-us	1	0
2	2	Topic	checkoutasguestorregister	1	0
3	3	Topic	conditions-of-use	1	0
4	4	Topic	contactus-2	1	0
5	5	Topic	forums	1	0
6	6	Topic	welcome-to-our-store	1	0
7	7	Topic	about-login-registration	1	0
8	8	Topic	privacy-notice	1	0
9	9	Topic	pagenotfound	1	0
10	10	Topic	shipping-returns	1	0
11	11	Topic	applyvendor	1	0
12	12	Topic	vendortermsofservice	1	0
13	1	BaseEntity	painted-plinth	0	0
14	1	Product	painted-plinth-2	1	0
15	17	Category	painted	1	0
16	1	BaseEntity	plinth	1	0
17	2	BaseEntity	painted-2	0	0
18	3	BaseEntity	30x30cm	0	0
19	4	BaseEntity	70cm-high	1	0
20	1	ProductTag	plinth-2	1	0
21	25	Category	plinths	1	0
22	30	BaseEntity	naked	1	0
23	18	Category	hardwood	1	0
24	31	BaseEntity	steel-frame	1	0
25	32	BaseEntity	steel-sheet	1	0
26	33	BaseEntity	acrylic	1	0
27	33	Category	acrylic-2	1	0
28	34	BaseEntity	concrete	1	0
29	35	BaseEntity	curved	1	0
30	20	Category	plinth-accessories	1	0
31	36	BaseEntity	acrylic-covers	1	0
32	37	BaseEntity	display-cases	1	0
33	22	Category	furniture	1	0
34	38	BaseEntity	lectern-book-plinth	1	0
35	39	BaseEntity	collection-box	1	0
36	40	BaseEntity	gallery-bench	1	0
37	41	BaseEntity	folding-gallery-bench	1	0
38	42	BaseEntity	gallery-table	1	0
39	43	BaseEntity	folding-gallery-table	1	0
40	44	BaseEntity	front-of-house-desk	1	0
41	45	BaseEntity	folding-display-unit	1	0
42	46	BaseEntity	simple-desk	1	0
43	47	BaseEntity	mobile-gallery-wall	1	0
44	48	BaseEntity	gallery-stool	1	0
45	49	BaseEntity	trestle	1	0
46	2	BaseEntity	painted-plinth-30x30x40	1	0
47	2	Product	painted-plinth-30x30x40-2	1	0
48	43	Category	folding-gallery-table-2	1	0
49	26	Category	sales-department	1	0
50	30	Category	naked-2	1	0
51	2	ProductTag	painted-3	1	0
52	3	ProductTag	30x30cm-2	1	0
53	5	BaseEntity	40cm-tall	1	0
54	6	BaseEntity	60cm-tall	1	0
55	50	BaseEntity	test-category	1	0
56	3	BaseEntity	painted-plinth-30x30x50	1	0
57	3	Product	painted-plinth-30x30x50-2	1	0
58	7	BaseEntity	50cm-tall	1	0
```
Note that there are entries with EntityName of BaseEntity that were created as Category. Tags seem randomy assigned BaseEntity.

I cleared the entries in this table and recreated them by manually editing categories, then products. Editing products generated the error described in the issue topic, so I cleared the tags from each product. I finally added topics to produce this table:
```
Id	EntityId	EntityName	Slug	IsActive	LanguageId
60	26	Category	sales-department	1	0
61	25	Category	plinths	1	0
62	50	Category	test-category	1	0
63	17	Category	painted	1	0
64	18	Category	hardwood	1	0
65	31	Category	steel-frame	1	0
66	32	Category	steel-sheet	1	0
67	33	Category	acrylic	1	0
68	34	Category	concrete	1	0
69	35	Category	curved	1	0
70	30	Category	naked	1	0
71	20	Category	plinth-accessories	1	0
72	36	Category	acrylic-covers	1	0
73	37	Category	display-cases	1	0
74	22	Category	furniture	1	0
75	38	Category	lectern-book-plinth	1	0
76	39	Category	collection-box	1	0
77	40	Category	gallery-bench	1	0
78	41	Category	folding-gallery-bench	1	0
79	42	Category	gallery-table	1	0
80	44	Category	front-of-house-desk	1	0
81	45	Category	folding-display-unit	1	0
82	46	Category	simple-desk	1	0
83	47	Category	mobile-gallery-wall	1	0
84	48	Category	gallery-stool	1	0
85	49	Category	trestle	1	0
86	27	Category	hire-department	1	0
87	28	Category	hire-plinths	1	0
88	29	Category	hire-painted	1	0
89	2	Product	painted-plinth-30x30x40	1	0
90	3	Product	painted-plinth-30x30x50	1	0
91	1	Product	painted-plinth-30x30x60	1	0
95	11	Topic	applyvendor	1	0
96	2	Topic	checkoutasguestorregister	1	0
98	4	Topic	contact-us	1	0
99	5	Topic	forums	1	0
100	6	Topic	home	1	0
101	7	Topic	about-login-registration	1	0
102	9	Topic	pagenotfound	1	0
103	12	Topic	vendortermsofservice	1	0
104	10	Topic	shipping-returns	1	0
105	8	Topic	privacy-notice	1	0
106	3	Topic	conditions-of-use	1	0
107	1	Topic	about-us	1	0
```

Back to the tags issue, I edited the product tags which introduced new entries to the table:
```
Id	EntityId	EntityName	Slug	IsActive	LanguageId
108	3	ProductTag	30x30cm	1	0
109	5	ProductTag	40cm-tall	1	0
110	7	ProductTag	50cm-tall	1	0
111	6	ProductTag	60cm-tall	1	0
112	4	ProductTag	70cm-tall	1	0
113	2	ProductTag	painted-2	1	0
114	1	ProductTag	plinth	1	0
```

I deleted the product tags using Catalog > Product Tags. This neither deleted the entries in the table nor set the flag IsActive to false.

I removed the Tags from the table directly.

Back to Products, I edited the first product and re-inserted the tags required. This inserted the following rows to the table:
```
Id	EntityId	EntityName	Slug	IsActive	LanguageId
115	8	BaseEntity	plinth	1	0
116	9	BaseEntity	painted-2	1	0
117	10	BaseEntity	30x30cm	1	0
118	11	BaseEntity	40cm-high	1	0
```
Hmm, not sure this is correct... I edited these tags tags via the Product Tags page, resulting in new entries in the table in addition to those created in the previous step.
```
Id	EntityId	EntityName	Slug	IsActive	LanguageId
120	12	BaseEntity	plinth	1	0
121	13	BaseEntity	painted-2	1	0
122	14	BaseEntity	30x30cm	1	0
123	15	BaseEntity	40cm-high	1	0
124	14	ProductTag	30x30cm-2	1	0
125	15	ProductTag	40cm-high-2	1	0
126	13	ProductTag	painted-3	1	0
127	12	ProductTag	plinth-2	1	0
```

I hope this helps track down the bug that looks to be assigning the wrong entity type when creating tags via product page editing or duplication.

I'd like to know whether the system is designed to delete unused entries in the table - by not clearing them out on changes the result is some pretty ugly friendly urls being created.

If you look at the table at the starting point of this post, there could well be other methods not working as expected. Half of my links were not working.

The whole system of generating friendly urls should be checked for operation as I am sure that there is a larger underlying problem.

**Comment by AndreiMaz** [comment]
Thanks. But the second issue is already fixed

P.S. Please note that 4.10 is not released yet. And not recommended to use in production

**Comment by EddieGreen** [comment]
Returning to the original issue, I discovered that when updating a product and adding a new tag, this generated the error. By deleting all entries in the **Product_ProductTag_Mapping** table for the product prior to saving the edit, the error did not occur. I further discovered that even if no edits were made to tag entries, saving a product edit generated the error for any product which contained tags.

I believe this error is caused by not deleting all entries in the Product_ProductTag_Mapping table for the product prior to saving current entries, or alternatively checking the contents of this table for changes to tag entries before writing new ones to the table, or by constructing the table with primary keys on both fields.

My observations regarding the creation and editing of friendly urls seem unconnected and this probably should be moved to a separate issue

I'd like to contribute to the development of NopCommerce, but have never been through the process of contributing via GitHub. Is there anyone in the UK already contributing that could help me set up, maybe share a screen?

**Comment by EddieGreen** [comment]
> Thanks. But the second issue is already fixed

Can you provide link to issue please. this does not appear in dev branch.

> Please note that 4.10 is not released yet. And not recommended to use in production

Thanks for the advice, I'm not in production, road testing the GDPR additions

**Comment by skoshelev** [comment]
@EddieGreen 

> Can you provide link to issue please. this does not appear in dev branch.

see this commit 332cde6

**Comment by skoshelev** [comment]
@EddieGreen 

> When a product with tags is copied, further editing generates an error (see log below) and returns a blank screen.

Can you reproduce this issue with the last develop branch changes, because I can't, if so, please  provide more details to reproduce it.

**Comment by skoshelev** [comment]
Can't reproduce

Closed #3040
